File: previous_solution/greedy.cpp
#include <iostream>
#include <algorithm>
#include <limits.h>
#include <utility>
#include <vector>
#include "Useful.cpp"

#define p 25
using namespace std;



vector<bloque> A;
vector<bloque> B;

double peso(vector<Tupla> &matchings){
    double suma_total=0;
    double tempa = A[matchings[0].first].longitud;
    double tempb = B[matchings[0].second].longitud;
    double t_a = matchings[0].first;
    double t_b = matchings[0].second;
    for (int i = 1 ; i < matchings.size(); i++){
        if(matchings[i].first == t_a){
            tempb += B[matchings[i].second].longitud;
        }
        else if (matchings[i].second == t_b){
            tempa += A[matchings[i].first].longitud;
        }
        else{
            suma_total += tempa/tempb;
            t_a = matchings[i].first;
            t_b = matchings[i].second;
            tempa = A[matchings[i].first].longitud;
            tempb = B[matchings[i].second].longitud;
        }
    }
    suma_total += tempa/tempb;
    return suma_total;
}

vector<Tupla> greedy_min() {
    int i, j;
    vector<Tupla> matchings;
    i = 0;
    j = 0;
    int cont_B = 0;
    int cont_A = 0;
    while (i < A.size() - 1 && j < B.size() - 1) {
        if (B[j + 1].longitud < B[j].longitud) {
            if (cont_B) {
                j++;
                i++;
                cont_B = 0;
            } else {
                matchings.emplace_back(Tupla(i, j));
                j++;
                cont_A++;
                if(j == B.size() - 1)
                    i++;
            }
        } else if (A[i].longitud < A[i + 1].longitud) {
            matchings.emplace_back(Tupla(i, j));
            if(cont_A){
                i++;
                j++;
                cont_B = 0;
                cont_A = 0;
            }else{
                i++;
                cont_B++;
                if(i == A.size()-1)
                    j++;
            }
        }
        else{
            matchings.emplace_back(Tupla(i, j));
            i++;
            j++;
            cont_B = 0;
            cont_A = 0;
        }
    }
    if (i == A.size() - 1) {
        while (j < B.size()) {
            matchings.emplace_back(Tupla(i, j));
            j++;
        }
    } else {
        while (i < A.size()) {
            matchings.emplace_back(Tupla(i, j));
            i++;
        }
    }
        return matchings;
    }
    
    
    pair<vector<Tupla>, double> MIN_MATCHING(vector<int> a, vector<int> b ) {
        ObtenerBloques(A,a);
        ObtenerBloques(B,b);
        pair<vector<Tupla>, double> ret1;
        ret1.first = greedy_min();
        ret1.second = peso(ret1.first);
        return ret1;
    }


int main() {

  //  int a[p]={0, 1 , 0 , 0 , 1 , 1, 0 , 1 , 0/* , 1  , 1 , 1 , 0 , 1 , 1 , 0 , 1 */};
  //  int b[p]={0, 0,  1 , 1 , 0 , 1 , 1, 0 , 1/* , 0  , 1 , 1 , 0 , 0 , 0 , 0 , 0 */};
    vector<int> a; // = {0, 1 , 0 , 0 , 1 , 1, 0 , 1 , 0 };
    vector<int> b; // = {0, 0,  1 , 1 , 0 , 1 , 1, 0 , 1 };
    Cargar_consola(a,b);

    auto result = MIN_MATCHING(a,b);

    for(auto v : result.first){
        cout<<"("<<v.first<<","<<v.second<<") ";
    }
    cout << endl << result.second;
    return 0;
}

File: 2da Entrega/Algorithms/Greedy.py
import sys
sys.path.append("../")
import Useful as us

A = []
B = []
matchings = []

def peso():
    global A, B
    suma_total=0
    if len(matchings)==0:
        return 0
    t_a = matchings[0][0]
    t_b = matchings[0][1]
    tempa = A[t_a].longitud
    tempb = B[t_b].longitud
    for i in range  (1,len(matchings),1):
        if(matchings[i][0] == t_a):
            tempb += B[matchings[i][1]].longitud
        
        elif (matchings[i][1] == t_b):
            tempa += A[matchings[i][0]].longitud
        else:
            suma_total += tempa/tempb
            t_a = matchings[i][0]
            t_b = matchings[i][1]
            tempa = A[matchings[i][0]].longitud
            tempb = B[matchings[i][1]].longitud
    suma_total += tempa/tempb
    return suma_total

def greedy_min():
    global matchings
    i = 0
    j = 0
    cont_B = 0
    cont_A = 0
    while (i < len(A) - 1 and j < len(B) - 1):
        if (B[j + 1].longitud < B[j].longitud):
            if (cont_B):
                j += 1
                i += 1
                cont_B = 0
            else:
                matchings.append([i,j])
                j += 1
                cont_A += 1
                if( j == len(B)- 1 ):
                    i += 1  
        elif (A[i].longitud < A[i + 1].longitud):
            matchings.append([i,j])
            if(cont_A):
                i += 1
                j += 1 
                cont_B = 0
                cont_A = 0
            else:
                i += 1
                cont_B += 1 
                if(i == len(A)-1):
                    j += 1
        else:
            matchings.append([i,j])
            i += 1 
            j += 1 
            cont_B = 0
            cont_A = 0

    if (i == len(A) - 1):
        while (j < len(B)):
            matchings.append([i,j])
            j += 1
    else:
        while (i < len(A)):
            matchings.append([i,j])
            i += 1

def MIN_MATCHING(a, b, GetSubmatching = False):
    global A, B
    A = us.ObtenerBloques(a)
    B = us.ObtenerBloques(b)
    matchings.clear()
    greedy_min()
    if(GetSubmatching):
        return us.GetSubMatchings(matchings, A, B)
    return peso()